Add support for Sensirion SEN5x environmental sensor

anbreww opened the pull request:
Adds support for the [Sensirion SEN5x series environmental sensors](https://sensirion.com/products/catalog/SEN54/) (so far only tested on the SEN54 since it's the only one that's been released)

Supports reading of PM1.0, PM2.5, PM4.0, PM10.0, Humidity, Temperature, and VOC index. 
NOx will be available in the SEN55 but it's not yet implemented in this component.

# What does this implement/fix?

Quick description and explanation of changes

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

(TODO : documentation page)

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:
```yaml
sensor:
  - platform: sen5x
    pm_1_0:
      name: "Bedroom <1µm Weight concentration"
    pm_2_5:
      name: "Bedroom <2.5µm Weight concentration"
    pm_4_0:
      name: "Bedroom <4µm Weight concentration"
    pm_10_0:
      name: "Bedroom <10µm Weight concentration"
    tvoc:
      name: "Bedroom VOC Index"
    humidity:
      name: "Bedroom Humidity"
    temperature:
      name: "Bedroom Temperature"
    update_interval: 30s
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

anbreww committed:
add support for Sensirion SEN5x sensors

Supports reading of PM1.0, PM2.5, PM4.0, PM10.0, Humidity, Temperature,
and VOC index. NOx will be available in the SEN55 but it's not yet
implemented in this component.

anbreww committed:
update codeowners

anbreww committed:
fix clang formatting

anbreww committed:
add required delays between readings

jesserockz commented on esphome/components/sen5x/sen5x.cpp:
```diff
@@ -0,0 +1,286 @@
+
+static const char *const TAG = "sen5x";
+
+static const uint16_t SEN5x_CMD_START_MEASUREMENT = 0x0021;
+static const uint16_t SEN5x_CMD_START_MEASUREMENT_RHT_GAS_ONLY = 0x0037;
+static const uint16_t SEN5x_CMD_STOP_MEASUREMENT = 0x0104;
+static const uint16_t SEN5x_CMD_READ_DATA_READY_FLAG = 0x0202;
+static const uint16_t SEN5x_CMD_READ_MEASURED_VALUES = 0x03C4;
+static const uint16_t SEN5x_CMD_READ_WRITE_T_COMPENSATION = 0x60B2;
+static const uint16_t SEN5x_CMD_READ_WRITE_WARM_START_PARAMS = 0x60C6;
+static const uint16_t SEN5x_CMD_READ_WRITE_VOC_ALGO_TUNING_PARAMS = 0x60D0;
+static const uint16_t SEN5x_CMD_READ_WRITE_NOX_ALGO_TUNING_PARAMS = 0x60E1;
+static const uint16_t SEN5x_CMD_READ_WRITE_RHT_ACCEL_MODE = 0x60F7;
+static const uint16_t SEN5x_CMD_READ_WRITE_VOC_ALGO_STATE = 0x6181;
+static const uint16_t SEN5x_CMD_START_FAN_CLEANING = 0x5607;
+static const uint16_t SEN5x_CMD_READ_WRITE_AUTO_CLEANING_INTVL = 0x8004;
+static const uint16_t SEN5x_CMD_READ_PRODUCT_NAME = 0xD014;
+static const uint16_t SEN5x_CMD_READ_SERIAL_NUMBER = 0xD033;
+static const uint16_t SEN5x_CMD_READ_FIRMWARE_VERSION = 0xD100;
+static const uint16_t SEN5x_CMD_READ_DEVICE_STATUS = 0xD206;
+static const uint16_t SEN5x_CMD_CLEAR_DEVICE_STATUS = 0xD210;
+static const uint16_t SEN5x_CMD_RESET = 0xD304;
```
Need to make the `x` captilised on all of these too

anbreww replied:
Thanks for taking the time to review my code! I'll fix this in the next commit.

jesserockz commented on esphome/components/sen5x/sen5x.cpp:
```diff
@@ -0,0 +1,286 @@
+      return;
+    }
+
+    delay(20);
```
Does it really need a full 20ms delay here and below?

anbreww replied:
Probably not, but those are the values given in the datasheet. I've tried with no delay at all and the sensor doesn't respond, but haven't tried with less than 20 ms yet. Otherwise would you suggest using set_timeout instead so we don't stay blocked in the init function for too long?

jesserockz replied:
Its probably fine here. It wont actually be blocking init/setup because its already delayed 500ms into booting, so it will be a random loop down the line. Should be ok as it is only done once anyway

jesserockz commented on esphome/components/sen5x/sen5x.cpp:
```diff
@@ -0,0 +1,286 @@
+  }
+
+  this->set_timeout(50, [this]() {
+    uint16_t raw_data[20];
```
Why length 20 here when you only need 8?
```suggestion
    uint16_t raw_data[8];
```

anbreww replied:
Thanks, I missed that

anbreww reviewed (commented):
Just a heads up : even if these changes are approved, this PR isn't quite ready yet.
It's missing some sanity checks to avoid sending missing data (for the SEN50 for example), and it looks like the readings are off for about a minute after the sensor resets

anbreww committed:
Fix capitalization of SEN5X constants

anbreww committed:
sen5x: provisional support of NOx

Implemented missing functions to read NOx from the sen55 sensor
Handle invalid sensor data by sending NAN

anbreww committed:
fix formatting

jesserockz commented on esphome/components/sen5x/sen5x.cpp:
```diff
@@ -0,0 +1,292 @@
+    int16_t voc_index_int = static_cast<int16_t>(raw_data[6]);
+    int16_t nox_index_int = static_cast<int16_t>(raw_data[7]);
+
+
+    float pm_1_0 = pm_1_0_int == UINT_INVALID ? NAN : pm_1_0_int / 10.0f;
+    float pm_2_5 = pm_2_5_int == UINT_INVALID ? NAN : pm_2_5_int / 10.0f;
+    float pm_4_0 = pm_4_0_int == UINT_INVALID ? NAN : pm_4_0_int/ 10.0f;
+    float pm_10_0 = pm_10_0_int == UINT_INVALID ? NAN : pm_10_0_int/ 10.0f;
```
Just to fix CI
```suggestion
    float pm_1_0 = pm_1_0_int == UINT_INVALID ? NAN : pm_1_0_int / 10.0f;
    float pm_2_5 = pm_2_5_int == UINT_INVALID ? NAN : pm_2_5_int / 10.0f;
    float pm_4_0 = pm_4_0_int == UINT_INVALID ? NAN : pm_4_0_int / 10.0f;
    float pm_10_0 = pm_10_0_int == UINT_INVALID ? NAN : pm_10_0_int / 10.0f;
```

martgras commented:
Unfortunately I missed this PR when I wrote my implementation for a sen5x component.
I believe my version is more or less complete but not submitted because it's dependent on another PR I submitted.
My code is at https://github.com/martgras/esphome/tree/sen5x-support/esphome/components/sen5x
See my comment here https://github.com/esphome/feature-requests/issues/1689#issuecomment-1096387319
